Added Browser Extension Security Cheat Sheet

aakarshgopishetty opened the pull request:
### Summary:
This PR adds a new cheat sheet: **Top 10 Browser Extension Security Vulnerabilities** to help developers build safer browser extensions.

### Key Additions:
- Covers common vulnerabilities like **permissions overreach, XSS, insecure storage, and malicious updates**.
- Provides **code examples & mitigations** for each issue.
- Includes **security best practices from Chrome & Mozilla documentation**.

### References:
- **Google Chrome Extension Security Guide**: https://developer.chrome.com/docs/extensions/mv3/security/  
- **Mozilla Extension Security Best Practices**: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Security_best_practices  

This PR fixes issue #1516.  

Looking forward to feedback! 🚀

szh commented:
We recently merged a cheat sheet by almost the [same name](https://cheatsheetseries.owasp.org/cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.html). If this contains anything omitted by that one, please submit a PR to improve it.

aakarshgopishetty commented:
Thanks for the update! I reviewed the existing Browser Extension Vulnerabilities Cheat Sheet and noticed that it lacks detailed explanations, code examples, and mitigation strategies.  

Instead of adding a new file, I will update the existing cheat sheet with additional details to improve it. I’ll submit a new PR for this soon.

aakarshgopishetty commented:
> We recently merged a cheat sheet by almost the [same name](https://cheatsheetseries.owasp.org/cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.html). If this contains anything omitted by that one, please submit a PR to improve it.

Thanks for the feedback! I have updated this PR to **modify the existing Browser Extension Vulnerabilities Cheat Sheet** instead of adding a new one.  

### 🔹 Changes Made:
- **Replaced the previous cheat sheet content** with a more detailed version.
- **Added missing explanations, code examples, and mitigation steps** to improve clarity.  
- **Kept all key security concerns while enhancing readability.**  

Please review the updated version, and let me know if any further changes are needed. 🚀

szh commented on cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.md:
I'd recommend removing this. See https://cheatsheetseries.owasp.org/cheatsheets/Pinning_Cheat_Sheet.html#when-do-you-perform-pinning

```suggestion
```

szh commented on cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.md:
I recommend adding a link here to the CSP cheat sheet. Same for other topics that have dedicated cheat sheets.

szh commented on cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.md:
Better yet, don't fetch updates within the extension, and only rely on updates from the extension marketplace. See "Don’t inject or incorporate remote scripts" in the [Firefox docs](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Security_best_practices)

aakarshgopishetty commented:
Thank you for the review and feedback! 🙌

I have made the requested changes:
✅ Removed the certificate pinning reference as per the OWASP Pinning Cheat
Sheet.
✅ Added a link to the *CSP Cheat Sheet* for more details.
✅ Updated the mitigation for malicious updates to recommend relying on
*marketplace
updates* instead of fetching remote scripts.

Let me know if any further modifications are needed. Looking forward to
your review! 🚀

On Wed, Mar 19, 2025 at 6:54 PM Shlomo Zalman Heigh <
***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.md
> <https://github.com/OWASP/CheatSheetSeries/pull/1630#discussion_r2003309655>
> :
>
> >
> -Users may be unaware of how their data is being collected or used, leading to potential abuse of their information without consent or awareness.
> +🚨 Vulnerability
> +Some extensions send sensitive data over unsecured HTTP connections, making it vulnerable to interception by attackers.
> +
> +🔍 Example
> +
> +javascript
> +Copy code
> +fetch('http://example.com/api/data');
> +🔒 Mitigation
> +
> +Always use HTTPS for external communications to prevent data theft.
> +Where possible, implement certificate pinning to prevent man-in-the-middle (MITM) attacks.
>
> I'd recommend removing this. See
> https://cheatsheetseries.owasp.org/cheatsheets/Pinning_Cheat_Sheet.html#when-do-you-perform-pinning
> ⬇️ Suggested change
>
> -Where possible, implement certificate pinning to prevent man-in-the-middle (MITM) attacks.
>
> ------------------------------
>
> In cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.md
> <https://github.com/OWASP/CheatSheetSeries/pull/1630#discussion_r2003314390>
> :
>
> > +
> +🔍 Example
> +
> +json
> +Copy code
> +{
> +  "manifest_version": 2,
> +  "name": "My Extension",
> +  "content_security_policy": "default-src 'self'"
> +}
> +🔒 Mitigation
> +
> +Define a strict CSP in the extension’s manifest.json file.
> +Use nonce-based or hash-based policies to allow only trusted scripts.
> +Block execution of inline scripts and restrict third-party content sources.
> +
>
> I recommend adding a link here to the CSP cheat sheet. Same for other
> topics that have dedicated cheat sheets.
> ------------------------------
>
> In cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.md
> <https://github.com/OWASP/CheatSheetSeries/pull/1630#discussion_r2003320176>
> :
>
> > +🚨 Vulnerability
> +If an extension fetches updates from an untrusted server, an attacker could push malicious updates to all users.
> +
> +🔍 Example
> +
> +javascript
> +Copy code
> +chrome.runtime.onInstalled.addListener(() => {
> +  fetch('http://example.com/update-script.js')
> +    .then(response => response.text())
> +    .then(eval); // Unsafe!
> +});
> +🔒 Mitigation
> +
> +Sign extension updates with digital signatures to ensure authenticity.
> +Fetch updates only from a trusted HTTPS server.
>
> Better yet, don't fetch updates within the extension, and only rely on
> updates from the extension marketplace. See "Don’t inject or incorporate
> remote scripts" in the Firefox docs
> <https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Security_best_practices>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OWASP/CheatSheetSeries/pull/1630#pullrequestreview-2698404530>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BBH5GXCRASOFOJULC5VVC732VFVXZAVCNFSM6AAAAABZIWX66GVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMOJYGQYDINJTGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks & Regards

Aakarsh Gopishetty
***@***.***

aakarshgopishetty commented:
I have addressed all requested changes, including fixing Markdown lint errors. Please review and let me know if anything else is needed!

szh commented on cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.md:
```diff
@@ -38,7 +38,7 @@ chrome.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
     fetch('http://example.com/track', {
       method: 'POST',
-      body: JSON.stringify({ url: tab.url })
+      body: JSON.stringify({ URL: tab.url })
```
The reason this is being flagged is that it's not in a code element. See https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

szh commented:
I'm going to link again to the markdown guide... https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code
Please properly encapsulate code so it's formatted correctly.

aakarshgopishetty commented:
> I'm going to link again to the markdown guide... https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code
> Please properly encapsulate code so it's formatted correctly.

I sincerely apologize for the formatting mistakes and any inconvenience caused. 😞 I truly appreciate your patience and guidance throughout this process.

I've carefully reviewed the Markdown guidelines and have now:
✅ Properly formatted all code snippets.
✅ Fixed all Markdown linting issues.
✅ Improved readability by replacing icons with proper headings.

I’ve pushed the latest changes. Please let me know if anything else needs adjustment. Thanks again for your support and valuable feedback! 🙏

jmanico commented:
You have nothing to apologize for.
We are extremely grateful for your volunteerism and thank you for your efforts! Helping you through the process is our job.

Thank you again! 🙏🏼

szh commented on cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.md:
Here's an example of a code block in markdown:

```suggestion
```json
{
  ""manifest_version": 3,
  "name": "My Extension",
  "permissions": [
    "tabs",
    "http://*/*",
    "https://*/*",
    "storage"
  ]
}
```
```

szh commented on cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.md:
These top level headings should have "##", and the subheadings ("Example", etc) should have "###"

aakarshgopishetty commented:
Hi @szh,
I’ve made all the requested changes:
✅ Properly formatted JSON and other code blocks using Markdown syntax.
✅ Updated all headings to use the correct ## for top-level and ### for subheadings.
✅ Fixed all Markdown linting errors.

Please review the latest changes when you get a chance. Thanks again for your guidance!

szh commented on cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.md:
Extra quotes, causing the rest of the page to render improperly. Please use the Github (or another) markdown preview to review your work.
```suggestion
  "manifest_version": 3,
```

szh reviewed (changes requested):
Almost there...

aakarshgopishetty commented:
I sincerely appreciate you catching that mistake! I’m really sorry for the oversight—I’ve now fixed the JSON syntax and added the missing closing code block. Thanks for your patience, and I truly appreciate your time reviewing this. Please let me know if there’s anything else I need to adjust! 🙏

szh commented:
@aakarshgopishetty Please can you do the same for all the other code blocks. I'm repeating my request that you preview your changes in a Markdown viewer so you can see how they look. The amount of back and forth here is unnecessary - you should be able to fix basic formatting errors before you submit a PR. [For example](https://github.com/OWASP/CheatSheetSeries/blob/1053d85a175c425ccb8c998d37259fada44f044d/cheatsheets/Browser_Extension_Vulnerabilities_Cheat_Sheet.md#example-data-leakage).
The content you're submitting is good, but we can't accept contributions if they're not formatted properly 🤷‍♂️

szh commented:
Much better. I'm puzzled by the strange line breaks though - normally line breaks in markdown don't cause line breaks in the output. I think here there are trailing spaces that are causing it but I guess aren't being caught by the markdown linter?

![Screenshot 2025-03-25 at 2 51 58 PM](https://github.com/user-attachments/assets/a74b0038-0368-41e0-b271-4e9adf0f0dac)

szh commented:
Yeah, there are tons of trailing spaces

![Screenshot 2025-03-25 at 2 53 56 PM](https://github.com/user-attachments/assets/fd7ab629-694a-43ae-a8a9-47bd3cf19b4d)

I'll look into why the linter isn't flagging them.

szh commented:
Yeah, it's not flagging them because there are two spaces, which is supposed to be used to force a linebreak (br): https://github.com/updownpress/markdown-lint/blob/master/rules/009-no-trailing-spaces.md
There's no need for that here though, these line breaks should be removed.

aakarshgopishetty commented:
@szh Thank you for your patience and guidance! 

I have removed all trailing spaces and verified the formatting using GitHub's Markdown preview.

The Markdown now renders correctly with no unexpected line breaks.

I also double-checked that all code blocks and headings follow the OWASP formatting standards.

Everything should be good to go now! Please let me know if there's anything else needed, but I believe this is ready for merging. Appreciate all your help!

szh reviewed (approved):
LGTM! Woohoo!
